### PR TITLE
Rebase on bulk proposal

### PR DESCRIPTION
**Pull request by rossberg**
I had a hard time implementing the declarative element segments without the segment changes from the bulk proposal. So I decided to rebase this proposal on the bulk one.

That was a bit of a nightmare because there were many conflicts, duplicates and divergences. To make the combination work again I also had to

- add internal support for executing table.init for multiple tables (to combine active segment desugaring with multi tables),
- some tweaks and generalisations to handle element expressions now becoming regular constant expressions,
- fill in a few holes (e.g., missing entries in instruction index),
- fix a number of typos and bugs in the spec I encountered on the way.

However, I tried to keep the non-editorial changes to a minimum for this PR andI added various TODOs for follow-up consolidation steps:

- properly support multi-table table.copy and table.init (spec of execution, text format, tests),
- define a textual representation for element expressions consisting of more than one instruction,
- change table ops from bulk proposal to small-step semantics using table.get/set,
- fix element segment evaluation in spec,
- actually add declarative segments.

### CODE REVIEW

File: document/core/exec/instructions.rst [hunk 1]
```diff
@@ -689,6 +691,260 @@ Table Instructions
    \end{array}
 
 
+.. _exec-table.copy:
+
+:math:`\TABLECOPY`
+..................
+
+.. todo:: TODO: multi tables
```
Comment (lars-t-hansen) [inline document/core/exec/instructions.rst]: And also new bounds checking semantics, though I assume that that comes in via the bulk memory spec.

File: document/core/appendix/index-instructions.rst [hunk 2]
```diff
@@ -4,201 +4,201 @@
...
-(reserved)                              :math:`\hex{14}`                                                  
-(reserved)                              :math:`\hex{15}`                                                  
-(reserved)                              :math:`\hex{16}`                                                  
-(reserved)                              :math:`\hex{17}`                                                  
-(reserved)                              :math:`\hex{18}`                                                  
-(reserved)                              :math:`\hex{19}`                                                  
-:math:`\DROP`                           :math:`\hex{1A}`  :math:`[t] \to []`                          :ref:`validation <valid-drop>`            :ref:`execution <exec-drop>`
-:math:`\SELECT`                         :math:`\hex{1B}`  :math:`[t~t~\I32] \to [t]`                  :ref:`validation <valid-select>`          :ref:`execution <exec-select>`
-:math:`\SELECT~t`                       :math:`\hex{1C}`  :math:`[t~t~\I32] \to [t]`                  :ref:`validation <valid-select>`          :ref:`execution <exec-select>`
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/appendix/index-instructions.rst]: Doh, thanks! Merge conflicts on a markdown table are a massive pain, so easy to screw up.

File: test/core/elem.wast [hunk 3]
```diff
@@ -139,107 +192,105 @@
...
+  "out of bounds"
 )
 
-(assert_unlinkable
-  (module
-    (table 0 funcref)
-    (elem (i32.const 1))
-  )
-  "elements segment does not fit"
+;; Writing 0 elems outside of bounds is allowed now.
```
[the remark this answers is not in the document]
Comment (rossberg) [reply test/core/elem.wast]: Stale, removed.

File: test/core/linking.wast [hunk 4]
```diff
@@ -253,28 +253,30 @@
 )
 (assert_trap (invoke $Mt "call" (i32.const 7)) "uninitialized")
 
-(assert_unlinkable
+;; Unlike in the v1 spec, the elements stored before an out-of-bounds access
+;; persist after the instantiation failure.
```
[the remark this answers is not in the document]
Comment (rossberg) [reply test/core/linking.wast]: Adjusted comment.

File: document/core/exec/instructions.rst (hunk 1 above)
Comment (rossberg) [reply document/core/exec/instructions.rst]: Yes, I'll need to rebase again after this. :(

**Pull request merged**